File: CSharp/CubeSolver/Tools/BaseSolver.cs
using System.Collections;
using CubeSolver.Utils;
using NetCubeSolver;
using NetCubeSolver.ExactCover;

namespace CubeSolver.Tools;

public class BaseSolver
{
    protected List<int> MatrixColumnNames = new List<int>();
    protected List<BitArray> CoverMatrix = new List<BitArray>();

    protected Config CubeConfig;
    
    protected PuzzleSolution Solution;
    
    protected List<CoverNode> CoverNodes = new List<CoverNode>();

    protected CoverNode HeaderNode;

    protected void PrintPieceData()
    {
        foreach (var piece in CubeConfig.PuzzlePieces)
        {
            ProgramUtils.Write(GetPieceInfo(piece));
        }
    }
    
    protected string GetPieceInfo(PuzzlePiece piece)
    {
        return $"Piece {piece.Id}, using Symbol: {piece.Symbol}, has {piece.Components.Count} Cube(s) and {piece.MaxOrientation} Unique Orientation(s)";
    }
    
    protected void GenerateCoverMatrix()
    {
        MatrixColumnNames.Clear();
        CoverMatrix.Clear();

        foreach (var piece in CubeConfig.PuzzlePieces)
        {
            MatrixColumnNames.Add(piece.Id);
        }
        
        MatrixColumnNames.AddRange(Solution.GetPositionHeaders());

        for (int i = 0; i < CubeConfig.PuzzlePieces.Count; i++)
        {
            for (int j = 0; j < CubeConfig.PuzzlePieces[i].MaxOrientation; j++)
            {
                for (int z = 0; z < CubeConfig.CubeSize; z++)
                {
                    for (int y = 0; y < CubeConfig.CubeSize; y++)
                    {
                        for (int x = 0; x < CubeConfig.CubeSize; x++)
                        {
                            var row = GetMatrixRow(i, j, x, y, z);
                            if (row != null)
                            {
                                CoverMatrix.Add(row);
                            }
                        }
                    }
                }
            }
        }
    }

    protected void WriteToTextFile(Dictionary<string, PuzzleSolution> solutions, string[] keys)
    {
        using (var streamWriter =
               new StreamWriter(Path.Combine(ProgramUtils.OutputDirectory, $"{CubeConfig.OutputFileName}.txt")))
        {
            foreach (var piece in CubeConfig.PuzzlePieces)
            {
                streamWriter.WriteLine(GetPieceInfo(piece)); 
            }
            streamWriter.WriteLine();
            streamWriter.WriteLine();
            
            for (int i = 0; i < keys.Length; i++)
            {
                solutions[keys[i]].WriteToTextFile($"Solution {i}", streamWriter);
                streamWriter.WriteLine();
            }
            
        }
    }

    protected DancingLinks CreateDancingLinks()
    {
        return new DancingLinks(MatrixColumnNames, CoverMatrix);
    }

    private BitArray? GetMatrixRow(int pieceId, int orientation, int x, int y, int z)
    {
        var row = new BitArray(MatrixColumnNames.Count, false);
        row[pieceId] = true;
        var piece = CubeConfig.PuzzlePieces[pieceId];
        piece.CurrentOrientation = orientation;
        var transform = new SubPiece(x, y, z);
        piece.Transform(transform);
        foreach (var component in piece.Geometry)
        {
            var index = FindColumn(component);
            if (index == -1)
            {
                return null;
            }
            row[index] = true;
        }
        return row;
    }

    private int FindColumn(SubPiece subPiece)
    {
        if (subPiece.X < 0 || subPiece.X >= CubeConfig.CubeSize || subPiece.Y < 0 || subPiece.Y >= CubeConfig.CubeSize || subPiece.Z < 0 || subPiece.Z >= CubeConfig.CubeSize) return -1;
        
        return CubeConfig.PuzzlePieces.Count + subPiece.X + (subPiece.Y * CubeConfig.CubeSize) + (subPiece.Z * CubeConfig.CubeSize * CubeConfig.CubeSize); 
    }
}

File: Kotlin/app/src/main/kotlin/ca/jwolf/cubesolver/tools/BaseSolver.kt
package ca.jwolf.cubesolver.tools

import ca.jwolf.cubesolver.configs.PuzzleConfig
import ca.jwolf.cubesolver.configs.PuzzlePiece
import ca.jwolf.cubesolver.configs.SubPiece
import ca.jwolf.cubesolver.datamodels.PuzzleSolution
import ca.jwolf.cubesolver.utils.ProgramUtils
import java.io.File
import java.nio.file.Paths
import java.util.BitSet

abstract class BaseSolver {

    protected val matrixColumnNames = mutableListOf<Int>()
    protected val coverMatrix = mutableListOf<BitSet>()

    // inheriting classes must set these prior to it being useful
    protected lateinit var solution: PuzzleSolution
    protected lateinit var config: PuzzleConfig

    protected fun printPieceData()
    {
        config.puzzlePieces.forEach { piece ->
            ProgramUtils.writeLine(getPieceInfo(piece))
        }
    }



    protected fun generateCoverMatrix(){

        matrixColumnNames.clear()
        coverMatrix.clear()

        config.puzzlePieces.forEach { piece ->
            matrixColumnNames.add(piece.id)
        }

        matrixColumnNames.addAll(solution.getPositionalHeaders())

        for(i in 0 until config.puzzlePieces.size){
            for (o in 0 until config.puzzlePieces[i].maxOrientation){
                for (z in 0 until config.cubeSize){
                    for (y in 0 until config.cubeSize){
                        for (x in 0 until config.cubeSize){
                            val row = getMatrixRow(i, o, x, y, z)
                            if (row != null) {
                                coverMatrix.add(row)
                            }
                        }
                    }
                }
            }
        }
    }

    protected fun writeToTextFile(solutions: Map<String, PuzzleSolution>, sortedKeys: List<String>)
    {
        File(Paths.get(ProgramUtils.OutputDirectory, "${config.outputFileName}.txt").toString()).bufferedWriter().use { writer ->
            config.puzzlePieces.forEach { piece ->
                writer.write(getPieceInfo(piece))
                writer.newLine()
            }

            writer.newLine()
            writer.newLine()

            var solutionNumber = 0
            sortedKeys.forEach { key ->
                solutions[key]!!.writeToTextFile("Solution ${solutionNumber}", writer)
                solutionNumber++
                writer.newLine()
            }

        }
    }

    private fun getMatrixRow(pieceId: Int, orientationId: Int, x: Int, y: Int, z:Int): BitSet? {
        val row = BitSet(matrixColumnNames.size)
        row[pieceId] = true

        val piece = config.puzzlePieces[pieceId]
        piece.currentOrientation = orientationId
        val transform = SubPiece(x, y, z)
        piece.transform(transform)
        piece.geometry().forEach { component ->
            val index = findColumn(component)
            // piece would fall outside the valid area
            if (index == -1) {
                return null
            }
            row.set(index)
        }
        return row
    }

    private fun findColumn(subPiece: SubPiece): Int {
        if (subPiece.x < 0 || subPiece.x >= config.cubeSize ||
            subPiece.y < 0 || subPiece.y >= config.cubeSize ||
            subPiece.z < 0 || subPiece.z >= config.cubeSize)
        {
            return -1
        }
        return config.puzzlePieces.size + subPiece.x + (subPiece.y * config.cubeSize) + (subPiece.z * config.cubeSize * config.cubeSize)
    }

    private fun getPieceInfo(puzzlePiece: PuzzlePiece): String {
        return "Piece ${puzzlePiece.id}, using Symbol: ${puzzlePiece.symbol}, has ${puzzlePiece.components.size} Cube(s) and ${puzzlePiece.maxOrientation} Unique Orientation(s)"
    }

}